src/mlpack/core/optimizers/fw/update_span.hpp

```diff
@@ -0,0 +1,132 @@
+/**
+ * @file update_span.hpp
+ * @author Chenzhe Diao
+ *
+ * Update method for FrankWolfe algorithm, recalculate the optimal in the span
+ * of previous solution space. Used as UpdateRuleType.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_FW_UPDATE_SPAN_HPP
+#define MLPACK_CORE_OPTIMIZERS_FW_UPDATE_SPAN_HPP
+
+#include <mlpack/prereqs.hpp>
+#include "func_sq.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * Recalculate the optimal solution in the span of all previous solution space,
+ * used as update step for FrankWolfe algorithm.
+ *
+ * Currently only works for function in FuncSq class.
+ *
+ */
+class UpdateSpan
+{
+ public:
+  /**
+   * Construct the span update rule. The function to be optimized is 
+   * input here.
+   *
+   * @param function Function to be optimized in FrankWolfe algorithm.
+   */
+  UpdateSpan()
+  { /* Do nothing. */ }
+
+  /**
+   * Update rule for FrankWolfe, reoptimize in the span of current
+   * solution space.
+   *
+   * @param function function to be optimized.
+   * @param oldCoords previous solution coords, not used in this update rule.
+   * @param s current linearConstrSolution result.
+   * @param newCoords output new solution coords.
+   * @param numIter current iteration number.
+   */
+  void Update(FuncSq& function,
+              const arma::mat& oldCoords,
+              const arma::mat& s,
+              arma::mat& newCoords,
+              const size_t numIter)
+  {
+    // Add new atom into soluton space.
+    arma::uvec ind = find(s, 1);
+    arma::uword d = ind(0);
+    AddAtom(function, d);
+
+    // Reoptimize the solution in the current space.
+    arma::vec b = function.Vectorb();
+    arma::mat x = solve(currentAtoms, b);
+
+    // x has coords of only the current atoms, recover the solution
+    // to the original size.
+    newCoords = RecoverVector((function.MatrixA()).n_cols, x);
+  }
+
+  //! Get the current atom indices.
+  const arma::uvec& CurrentIndices() const { return currentIndices; }
+  //! Modify the current atom indices.
+  arma::uvec& CurrentIndices() { return currentIndices; }
+
+  //! Get the current atoms.
+  const arma::mat& CurrentAtoms() const { return currentAtoms; }
+  //! Modify the current atoms.
+  arma::mat& CurrentAtoms() { return currentAtoms; }
+
+ private:
+  //! Indices of the atoms in the current solution space.
+  arma::uvec currentIndices;
+
+  //! Current atoms in the solution space, ordered as currentIndices.
+  arma::mat currentAtoms;
+
+  //! Flag current indices is empty.
+  bool isEmpty = true;
+
+  //! Add atom into the solution space.
+  void AddAtom(FuncSq& function, const arma::uword k)
+  {
+    if (isEmpty)
```

Comment:
Can we use the armadillo function ``is_empty`` to do the same with the extra ``isEmpty`` parameter?